### ISSUE TITLE
Must input patch size be multiple of 1024?

### ISSUE THREAD

**Issue by GopiRajuMatta** [luost26/DMRDenoise#5]
Hello,

I need to use input patch size, which is <1024(let's say 512, 256).
While I am using input patch size of 512, this is popping up, "Input Error! The size of the point clouds should be a multiple of 1024. error in nnd get grad: invalid configuration argument" but training is still going on.

Can you please suggest me to handle point clouds which have patch size<1024.

Also, it is taking point cloud normals as input, but I didn't see in the code using normals while training.

Thank you
Gopi

**Comment by GopiRajuMatta** [comment]
Also,

I didn't change anything in the code, I am just simply retraining the existing one. But error is not going down. I ran it about 200 epochs, still training error is more or less same as of the start. It seems like just oscillating. What would be the reason?

Thank you
Gopi

**Comment by luost26** [comment]
Hi Gopi,

I am so sorry for replying so late. I am busy with some conference deadline these days and missed your messages somehow.

The limit of 1024 point is from the EMD loss operator. The EMD operator is directly taken from [this repo](https://github.com/Colin97/MSN-Point-Cloud-Completion/tree/master/emd).

If you want to train with point clouds whose patch size is less than 1024, I would recommend using Chamfer Distance (CD) loss, which doesn't have such limit.

[These two flags](https://github.com/luost26/DMRDenoise/blob/290a9e7a62b0fc2bd0ee2b906df3d353edf0eef9/models/denoise.py#L180) control which loss function to use during training.
To train a model using CD loss, you may simply add these two flags: `--loss_rec cd --loss_ds cd`

Thanks!

**Comment by luost26** [comment]
> Also, it is taking point cloud normals as input, but I didn't see in the code using normals while training.

Both training and denoising do not use normals. 
Can you specify which lines of code use normals ?

**Comment by luost26** [comment]
> I didn't change anything in the code, I am just simply retraining the existing one. But error is not going down. I ran it about 200 epochs, still training error is more or less same as of the start. It seems like just oscillating. What would be the reason?

Can you provide some details about the error ?

I guess in most cases it is caused by the EMD operator which is unstable. Some other people also reported that the EMD operator is sometimes problematic. On my GPU server, after I updated my CUDA driver to version 11, the EMD operator doesn't work anymore. 

Thus, I would recommend not using the EMD loss anymore if you have encountered errors from it. 
Try using CD loss instead. The CD loss only uses native Pytorch operator which is more reliable.

**Comment by GopiRajuMatta** [comment]
Thank you for the response. 

1) Yes, I will try using CD loss instead of EMD.
2) I am also saying, code doesn't use normals, but it is taking input normals also and some of the code is there to store normals and pass normals to the loss function. But only in loss function, normals are not used.
3) This is the information about error.
![1](https://user-images.githubusercontent.com/42478399/107184117-fa339f80-6a05-11eb-8512-15440830d6e8.png)
I changed seed value, and then it worked out. Does seed matter that much?

**Comment by luost26** [comment]
> I am also saying, code doesn't use normals, but it is taking input normals also and some of the code is there to store normals and pass normals to the loss function. But only in loss function, normals are not used.

I see. This is because I tried some [loss functions that take normals as input](https://arxiv.org/abs/2002.05968) when I was doing preliminary experiments. I did not use such loss functions in the final version but the interface that takes normals as input was not removed.


> I changed seed value, and then it worked out. Does seed matter that much?

I did encounter some cases in various projects (point cloud denoising, point cloud generation, graph neural networks, etc.) where some "bad" seeds lead to NaN during training.
I am also confused by this but I just tried different seeds as you did.
